sklearn/mixture/base.py

```diff
@@ -0,0 +1,425 @@
+                                   random_state=random_state).fit(X).labels_
+            resp[range(n_samples), label] = 1
+        elif self.init_params == 'random':
+            resp = random_state.rand(X.shape[0], self.n_components)
+            resp = resp / resp.sum(axis=1)[:, np.newaxis]
+        else:
+            raise ValueError("Unimplemented initialization method '%s'"
+                             % self.init_params)
+
+        self._initialize(X, resp)
+
+    @abstractmethod
+    def _initialize(self, X, resp):
+        """Initialize the model parameters of the derived class.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+
+        resp : array-like, shape = (n_samples, n_components)
+        """
+        pass
+
+    def fit(self, X, y=None):
+        """Estimate model parameters with the EM algorithm.
+
+        The method fit the model `n_init` times and set the parameters with
+        which the model has the largest likelihood or lower bound. Within each
+        trial, the method iterates between E-step and M-step for `max_iter`
+        times until the change of likelihood or lower bound is sless than
+        `tol`, otherwise, a `ConvergenceWarning` is raised.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            List of n_features-dimensional data points.  Each row
+            corresponds to a single data point.
+
+        Returns
+        -------
+        self
+        """
+        X = _check_X(X, self.n_components)
+        self.n_features = X.shape[1]
+        self._check_parameters()
+
+        max_log_likelihood = -np.infty
+
+        # if we enable warm_start, we will have a unique init
+        do_init = not(self.warm_start and self.initialized_)
+        n_init = self.n_init if do_init else 1
+
+        for init in range(n_init):
+            # print verbose msg
+            self._print_verbose_msg_init_beg(init)
+
+            if do_init:
+                self._initialize_parameters(X)
+                self.initialized_ = True
+
+            current_log_likelihood = -np.infty
+            self.converged_ = False
+
+            for n_iter in range(self.max_iter):
+                prev_log_likelihood = current_log_likelihood
+
+                current_log_likelihood, resp = self._e_step(X)
+                change = current_log_likelihood - prev_log_likelihood
+                self._print_verbose_msg_iter_end(n_iter, change)
+
+                if abs(change) < self.tol:
+                    self.converged_ = True
+                    break
+
+                self._m_step(X, resp)
+
+            if not self.converged_:
+                # compute the log-likelihood of the last m-step
+                warnings.warn('Initialization %d is not converged. '
+                              'Try different init parameters, '
+                              'or increase n_init, tol '
```

Comment:
User should check for `max_iter` rather than `n_init`: we are inside the `n_init` loop here.